ROADMAP.md

```diff
@@ -0,0 +1,155 @@
+
+### B. 🥊 Decentralized Hole Punching
+<!--- TODO: Link to GitHub Epic -->
+**Why**:  P2P networks can have a combination of both public and private nodes. While private nodes can dial nodes on the public Internet, they are unreachable from the outside as they are behind a NAT or a firewall. We need a mechanism to dial them. A [previous DHT crawl found that almost 63%](https://github.com/libp2p/specs/blob/master/ROADMAP.md#-hole-punching-on-tcp-and-quic) of the network was undialable.  This project aims to implement Decentralized Hole Punching in js-libp2p and bring it to parity with the Go and Rust implementations.
+
+**Goal**:
+#### 1. [AutoNat](https://github.com/libp2p/js-libp2p/issues/1005)
+Determine whether a node is public or private (located behind a firewall or a NAT.) This is a dependency for enabling the DHT in server mode by default for projects like js-ipfs.
+#### 2. [Circuit Relay v2](https://github.com/libp2p/js-libp2p/issues/1029)
+Connect to, request reservations, and establish a secure relay connection through discovered public relay node.
+#### 3. Add QUIC Transport
+Hole punching is [more reliable with UDP](https://www.notion.so/pl-strflt/NAT-Hole-punching-Success-Rate-2022-09-29-Data-Analysis-8e72705ca3cc49ab983bc5e8792e3e98#5b76991da8d24736abd486aa93e85a97) (therefore QUIC) than TCP. This requires adding support for QUIC in js-libp2p. There is some work [being done here](https://github.com/nodejs/node/pull/44325) to add support in node.js which we depend on.
+#### 4. Hole Punching
+Use [DCUtR](https://github.com/libp2p/specs/blob/master/relay/DCUtR.md) to synchronize hole punching
+
+### C. 🔮 Future-proof Observability
+***Why*** We release new versions of js-libp2p and we should analyze/profile each version for regressions. Furthrmore, we already expose [per-component metrics](https://github.com/libp2p/js-libp2p/issues/1060) in js-libp2p, we can improve this by adding introspection.
+
+***Goal***
+For each js-libp2p release, we ensure transfer performance does not degrade. We can observe/get introspective data for a running js-libp2p instance.
+Introduce an introspection protocol to js-libp2p, so we can hook into the [observer-toolkit](https://github.com/libp2p/observer-toolkit).
+
+#### 1. Benchmark transfer performance in CI
+Add [js-libp2p-transfer-performance](https://github.com/ipfs-shipyard/js-libp2p-transfer-performance) to a CI job to benchmark transfer times across releases.
+#### 2. [Add introspection to js-libp2p](https://github.com/libp2p/js-libp2p-daemon/pull/45)
+Add an introspection protocol for js-libp2p such that users of js-ipfs can utilize observer-toolkit to visualize introspection data.
```

Comment:
Not critical for merging this, but I don't understand what this gives us verses grafana dashboards.

Reply:
Related but Chainsafe are very keen on better metrics for Lodestar and we're kind of at the limit of what we can do with the current metrics implementation.  An overhaul to make it a bit more flexible could take the place of this - it'd be quite low-effort I think.

Reply:
> An overhaul to make it a bit more flexible could take the place of this - it'd be quite low-effort I think.

👍
This sounds good to me and in-line with the triage meeting discussions yesterday, I'll update this accordingly. Low-effort and high value is always good to have